### PR TITLE
#167727727 Implement email & password authentication

### PR DESCRIPTION
**Pull request by jsbuddy**
#### What this PR does
This PR implements the functionality for email and password authentication

#### Tasks completed
- create signin route
- create and implement signin controller
- create helper functions in user's service
- write tests for sign in route

#### How can this be manually tested
- Clone the repo
- Install dependencies, setup database and run `npm start:dev`
- Using postman or any http request tool, make a post request to `api/v1/user/signin`

#### What are the relevant pivotal tracker stories?
[#167727727](https://www.pivotaltracker.com/story/show/167727727)

### CODE REVIEW

**Comment by tejiri4** [comment]
Kindly Write test for all file changes you made nice work

File: src/routes/api/user.js [hunk 1]
```diff
@@ -49,5 +50,44 @@ const userRoute = (router) => {
...
+   *     requestBody:
+   *      description: User data object
+   *      required: true
+   *      content:
+   *       application/json:
+   *          schema:
+   *            $ref: '#/components/schemas/SignIn'
+   *     responses:
+   *       200:
+   *         description: Authentication successful
```
Comment (jsbuddy) [inline src/routes/api/user.js]: Include token in description

File: src/controllers/userController.js [hunk 2]
```diff
@@ -43,6 +44,29 @@ const signUp = async (req, res) => {
...
+ * @param {Object} req - server request
+ * @param {Object} res - server response
+ * @returns {Object} - custom response
+*/
+
+const signIn = async (req, res) => {
+  try {
+    const { email, password } = req.body;
+    const user = await findByEmail(email);
+    if (!user) return response(res, 404, 'error', { message: 'User not found, please check your email address' });
```
Comment (jsbuddy) [inline src/controllers/userController.js]: Extract response messages

**Comment by ladaposamuel** [comment]
Looks Good

**Review by ladaposamuel — APPROVED** [review]

**Review by tobslob — COMMENTED** [review]
LGTM

File: src/controllers/userController.js [hunk 3]
```diff
@@ -37,12 +39,33 @@ const signUp = async (req, res) => {
...
+    const { email, password } = req.body;
+    const user = await findByEmail(email);
+    if (!user) return response(res, 404, 'error', { message: messages.userNotFound });
+    const { password: hash, ...data } = user.dataValues;
+    const passwordsMatch = await comparePasswords(password, hash);
+    if (!passwordsMatch) return response(res, 400, 'error', { message: messages.incorrectPassword });
+    const token = authHelper.generateToken({ id: data.id });
+    return response(res, 200, 'success', { ...data, token });
+  } catch (error) {
+    response(res, 500, 'error', { error });
```
Comment (tejiri4) [inline src/controllers/userController.js]: Use sinon to test this line stub the User object and findOne method to throw error.

Comment (tejiri4) [reply src/controllers/userController.js]: this will make your test coverage pass

Comment (jsbuddy) [reply src/controllers/userController.js]: Okay @tejiri4,  looking into it

**Review by tejiri4 — APPROVED** [review]
LGTM

**Pull request merged**